Update new_horizon_ingest with the recent changes

bartekn opened the pull request.

ire-and-curses commented on services/horizon/internal/docs/plans/new_horizon_ingest.md:
```diff
@@ -97,54 +97,30 @@ Notes:
 - the `HistoryArchiveAdapter` supports both reading a ledger transaction result set via `GetLedger()` and reading ledger state via `GetState()`.
 - Both adapters support `GetLatestLedgerSequence()`, which allows a consumer to look up the most recent ledger information in the backend.
-- The adapters, rather than returning state and ledgers as objects stored in memory, return them as `Reader` objects. This is because the ledger state or a particular transaction state may not fit in memory, and must be processed as a stream.
+- The adapters, rather than returning state and ledgers as objects stored in memory, return them as `ReadCloser` objects. This is because the ledger state or a particular transaction state may not fit in memory, and must be processed as a stream.
```
Can we add a sentence with the rationale for adding `Close`?

ire-and-curses commented on services/horizon/internal/docs/plans/new_horizon_ingest.md:
```diff
@@ -97,54 +97,30 @@ Notes:
+#### Ingestion Pipeline
 
-At the center of ingestion is a filtering `Pipeline`, which is initialized with a series of filters for each kind of data that ingestion handles (ledger state, a full ledger update, and an archive ledger update). Each filter implements a function that reads data from a `Reader`, transforms it, and writes it to a `Writer`. A few example filters:
+At the center of ingestion is a `Pipeline`, which is initialized with a series of processors for each kind of data that ingestion handles (ledger state, a full ledger update, and an archive ledger update). Each processors implements a function that reads data from a `ReadCloser`, processes/filters it, and writes it to a `WriteCloser`. A few example processors:
```
Each processors implements -> Each processor implements

ire-and-curses commented on services/horizon/internal/docs/plans/new_horizon_ingest.md:
```diff
@@ -97,54 +97,30 @@ Notes:
-- I'd rename `Process` to `Processor` or even `Writer`
+- The processors form a tree and are processed from root to leaves.
+- Processing does not change the type of the data, but it can remove or change fields in the data
+- Processors can write/read artifacts (ex. average balances) in a `Store` that is shared between all processors in a pipeline. `Store` is ephemeral, it's contents is removed when pipeline processing is done.
```
it's contents is removed -> its contents are removed

ire-and-curses commented on services/horizon/internal/docs/plans/new_horizon_ingest.md:
```diff
@@ -154,6 +130,8 @@ The ingestion system can run as a stand-alone process or as part of a larger pro
 - `LiveSession`: ingest the latest ledgers as soon as the close. This is the standard operating mode for a live Horizon instance
 - `CheckpointsOnlySession`: ingest only history archive checkpoint data.
 
+Session are responsible for coordinating pipelines and ensuring data is in sync. When processors write to a single storage type, keeping data in sync is trivial. For example, for Postgres processors can share a single transaction object and commit when processing is done. However, there's also a significant challenge keeping data in sync when processors are writing to different storage types: any ...
```
Session are responsible -> Sessions are responsible

bartekn commented:
@ire-and-curses Thanks for the feedback!
> Looks good, just minor comments. Also, in `io.png`, `LedgerTransaction` should be updated to include TXindex info, and to replace `Transaction` with `TransactionEnvelope` as we discussed offline.

Updated but I removed `Transaction` prefix from all the fields (seems redundant as struct name is `LedgerTransaction`).

bartekn commented:
Merging because it's just a doc. Let me know if you see more errors and I'll submit a new PR.
